### PR TITLE
test(framework): print more helpful information to help debug db and k8s related flaky tests

### PR DESCRIPTION
**Pull request by jijiechen**
This PR adds two capabilities to our test framework:
1. Print database list and details from the db container when there is a postgre connection failure. There are flakes related to this were found recently.
2. Print pod events besides conditions when a pod is timeout to be available. This is helpful for diagnosing k8s related e2e cases.

### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues
  - N/A
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS 
  - Confirmed
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s)
  - N/A
  - Don't forget `ci/` labels to run additional/fewer tests
  - No need to add, will run normally
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? 
  - No need.
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) 
  - No need.

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/test/store/postgres/test_container.go [hunk 1]
```diff
@@ -108,7 +109,7 @@ func (v *PostgresContainer) Stop() error {
 
 type DbOption string
 
-func (v *PostgresContainer) Config(dbOpts ...DbOption) (*pg_config.PostgresStoreConfig, error) {
```
Comment (jijiechen) [inline pkg/test/store/postgres/test_container.go:111]: this parameter is now not used, so removed it.

**Comment by jijiechen** [comment]
Here is an example of database detail output:
https://github.com/jijiechen/kuma/actions/runs/7180097816/job/19554749026#step:6:45647

File: pkg/test/store/postgres/test_container.go [hunk 2]
```diff
@@ -153,3 +157,46 @@ func (v *PostgresContainer) Config(dbOpts ...DbOption) (*pg_config.PostgresStore
...
 	return cfg, nil
 }
+
+func (v *PostgresContainer) PrintDebugInfo(exportDbName string, expectedDbPort int) {
+	logEntries := make([]string, 0)
+	container := v.container
+	state, _ := container.State(context.TODO())
+	logEntries = append(logEntries, "db container details:")
+	logEntries = append(logEntries, fmt.Sprintf("trying to connect db %s on port %d, latest container: %s",
+		exportDbName, expectedDbPort, container.GetContainerID()))
```
Comment (michaelbeaumont) [inline pkg/test/store/postgres/test_container.go:167]: Kind of a nit but I find this function somewhat difficult to parse, maybe it's the style? I dunno

```suggestion
	container := v.container
	logEntries := string[]{
		"db container details:",
		fmt.Sprintf("trying to connect db %s on port %d, latest container: %s", exportDbName, expectedDbPort, container.GetContainerID()),		
	}
```

much easier to read

File: pkg/test/store/postgres/test_container.go [hunk 3]
```diff
@@ -153,3 +157,46 @@ func (v *PostgresContainer) Config(dbOpts ...DbOption) (*pg_config.PostgresStore
...
+
+	if state != nil {
```
Comment (michaelbeaumont) [inline pkg/test/store/postgres/test_container.go:169]:
```suggestion
	if state, _ := container.State(context.TODO()); state != nil {
```

generally just limiting the scope of variables as much as possible, couple more spots in the function.

File: test/framework/k8s_cluster.go [hunk 4]
```diff
@@ -1249,66 +1249,200 @@ func (c *K8sCluster) K8sVersionCompare(otherVersion string, baseMessage string)
...
+	deploy, err := k8s.GetDeploymentE(testingT, kubectlOptions, name)
+	if err != nil {
+		// might not be a Deployment, let's ignore it
+		return
+	}
+
+	deployDetails := deploymentDetails{Namespace: namespace,
+		Kind:          "Deployment",
+		objectDetails: newObjectDetails(name),
+		ReplicaSets:   make([]*objectDetails, 0),
```
Comment (michaelbeaumont) [inline test/framework/k8s_cluster.go:1268]: I think we can leave this out as the zero value. IMO the `make(.., 0)` pattern is noisy and never really needed.

File: test/framework/k8s_cluster.go [hunk 5]
```diff
@@ -1249,66 +1249,200 @@ func (c *K8sCluster) K8sVersionCompare(otherVersion string, baseMessage string)
...
-	podCond, _ := json.Marshal(pod.Status.Conditions)
-	podLogs, _ := k8s.GetPodLogsE(testingT, kubectlOptions, &pod, "")
-	if podLogs != "" {
-		logLines := strings.Split(podLogs, "\n")
-		if len(logLines) > 30 {
-			podLogs = strings.Join(logLines[len(logLines)-30:], "\n")
-		}
-		podLogs = "; last lines of pod logs: " + podLogs
+	pDetails := &podDetails{objectDetails: newObjectDetails(pod.Name)}
+	pDetails.Conditions = fromPodCondition(pod.Status.Conditions)
```
Comment (michaelbeaumont) [inline test/framework/k8s_cluster.go:1310]: Can we declare these all at once above when we initialize `pDetails`?

File: test/framework/k8s_cluster.go [hunk 6]
```diff
@@ -1249,66 +1249,200 @@ func (c *K8sCluster) K8sVersionCompare(otherVersion string, baseMessage string)
...
-		condition.Replicas = append(condition.Replicas, replicaSetCondition{Name: ars.Name, Conditions: ars.Status.Conditions})
+type objectDetails struct {
+	Name       string             `json:"name,omitempty"`
+	Conditions []*objectCondition `json:"conditions,omitempty"`
+	Events     []*simplifiedEvent `json:"events,omitempty"`
+}
+
+func newObjectDetails(name string) *objectDetails {
+	return &objectDetails{
+		name,
```
Comment (michaelbeaumont) [inline test/framework/k8s_cluster.go:1349]: named fields

File: pkg/test/store/postgres/test_container.go (hunk 2 above)
Comment (jijiechen) [reply pkg/test/store/postgres/test_container.go:167]: Thanks for the suggestions, I'll push a new commit and apply these.

File: test/framework/k8s_cluster.go (hunk 4 above)
Comment (jijiechen) [reply test/framework/k8s_cluster.go:1268]: Yeah, I'll remove them.

File: pkg/test/store/postgres/test_container.go (hunk 3 above)
Comment (jijiechen) [reply pkg/test/store/postgres/test_container.go]: Accepted!

File: test/framework/k8s_cluster.go (hunk 5 above)
Comment (jijiechen) [reply test/framework/k8s_cluster.go]: Sure, just updated code.

File: mk/k3d.mk [hunk 7]
```diff
@@ -113,9 +113,13 @@ k3d/configure/metallb:
 
 .PHONY: k3d/wait
 k3d/wait:
-	until \
```
Comment (jijiechen) [inline mk/k3d.mk:116]: The previous script will wait infinitely if the conditions will never met, which cause CI jobs never complete and we don't have a change to detect the detailed errors.

File: pkg/test/store/postgres/test_container.go [hunk 8]
```diff
@@ -153,3 +157,46 @@ func (v *PostgresContainer) Config(dbOpts ...DbOption) (*pg_config.PostgresStore
...
+
+	rDir := resourceDir()
+	clientKey, _ := os.ReadFile(path.Join(rDir, "certs/postgres.client.key"))
+	clientCert, _ := os.ReadFile(path.Join(rDir, "certs/postgres.client.crt"))
+	_ = container.CopyToContainer(context.Background(), clientKey, "/tmp/postgres.client.key", 0o600)
+	_ = container.CopyToContainer(context.Background(), clientCert, "/tmp/postgres.client.crt", 0o600)
+	listCmd := []string{"sh", "-c", "export PGSSLMODE=verify-full; export PGSSLROOTCERT=/var/lib/postgresql/rootCA.crt; " +
+		"export PGSSLCERT=/tmp/postgres.client.crt; export PGSSLKEY=/tmp/postgres.client.key; psql -h localhost --no-password -l -U kuma -P pager=off"}
+	exitCode, resultReader, err := container.Exec(context.Background(), listCmd)
+	if err == nil {
```
Comment (michaelbeaumont) [inline pkg/test/store/postgres/test_container.go:189]:
```suggestion
	if exitCode, resultReader, err := container.Exec(context.Background(), listCmd); err == nil {
```

File: pkg/test/store/postgres/test_container.go [hunk 9]
```diff
@@ -153,3 +157,46 @@ func (v *PostgresContainer) Config(dbOpts ...DbOption) (*pg_config.PostgresStore
...
+		fmt.Sprintf("trying to connect db %s on port %d, latest container: %s",
+			expectedDbName, expectedDbPort, container.GetContainerID()),
+	}
+
+	if state, _ := container.State(context.TODO()); state != nil {
+		logEntries = append(logEntries, fmt.Sprintf("container state: %s", state.Status))
+	}
+
+	logReader, _ := container.Logs(context.Background())
+	if logReader != nil {
```
Comment (michaelbeaumont) [inline pkg/test/store/postgres/test_container.go:174]:
```suggestion
	if logReader, _ := container.Logs(context.Background()); logReader != nil {
```

File: pkg/test/store/postgres/test_container.go [hunk 10]
```diff
@@ -153,3 +157,46 @@ func (v *PostgresContainer) Config(dbOpts ...DbOption) (*pg_config.PostgresStore
...
+		logs, _ := io.ReadAll(logReader)
+		if len(logs) > 0 {
```
Comment (michaelbeaumont) [inline pkg/test/store/postgres/test_container.go:176]:
```suggestion
		if logs, _ := io.ReadAll(logReader); len(logs) > 0 {
```

**Review by michaelbeaumont — APPROVED** [review]
Nice change, this will be helpful

File: mk/kind.mk [hunk 11]
```diff
@@ -85,10 +89,12 @@ kind/deploy/kuma: build/kumactl kind/load
 	@KUBECONFIG=$(KIND_KUBECONFIG) $(KUBECTL) wait --timeout=60s --for=condition=Available -n $(KUMA_NAMESPACE) deployment/kuma-control-plane
 	@KUBECONFIG=$(KIND_KUBECONFIG) $(KUBECTL) wait --timeout=60s --for=condition=Ready -n $(KUMA_NAMESPACE) pods -l app=kuma-control-plane
 	@KUBECONFIG=$(KIND_KUBECONFIG) $(KUBECTL) delete -n $(EXAMPLE_NAMESPACE) pod -l app=example-app
-	@until \
-      KUBECONFIG=$(KIND_KUBECONFIG) $(KUBECTL) wait -n kube-system --timeout=5s --for condition=Ready --all pods ; \
-    do \
-      echo "Waiting for the cluster to come up" && sleep 1; \
+	@TIMES_TRIED=0; \
+	MAX_ALLOWED_TRIES=20; \
+	until KUBECONFIG=$(KIND_KUBECONFIG) $(KUBECTL) wait -n kube-system --timeout=5s --for condition=Ready --all pods  || [[ $$TIMES_TRIED -ge $$MAX_ALLOWED_TRIES ]]; do \
```
[the remark this answers is not in the document]
Comment (jijiechen) [reply mk/kind.mk:94]: Good catch, it's not needed.

**Pull request merged**